tinygrad/tensor.py

```diff
@@ -3726,6 +3726,8 @@ def cast(self, dtype:DTypeLike) -> Tensor:
     if (dt:=to_dtype(dtype)) in {dtypes.uint8, dtypes.uint16} and dtypes.is_float(self.dtype):
       # NOTE: values within the int32 range and outside the unsigned dtype range will cause values to wrap around
       return self._apply_uop(UOp.cast, dtype=dtypes.int32)._apply_uop(UOp.cast, dtype=dt)
+    if self.dtype == dtypes.bfloat16 and dt == dtypes.float16:
```

Comment:
Can you use a rewrite rule to do this in the backends that require it? It doesn't belong here

Reply:
oh.. Any suggestion on how to find the correct "backend that requires it"? 

as for a test, I can't seem to find a way to replicate that exact
```c
*((half4*)((data0+(alu0+(gidx1<<14)+(lidx0<<11)+alu1)))) = make_half4(((half)(val0)),((half)(val1)),((half)(val2)),((half)(val3)));
```
that trigger the issue in a testcase either, so far I've tried lots of things that didn't work 🤔
I've tried several variations of https://gist.github.com/divinity76/587fb9721caecd7facb172d51926c14d 
without success, any suggestions?

Reply:
spent ~4 hours trying to reproduce it in a self-contained testcase without success. can't seem to trigger it without actually loading a real model 🤷‍♂️  giving up for now. if anyone has ideas, let me know.

Reply:
Fwiw @AlexCheema was able to reproduce this on a 4090: https://github.com/exo-explore/exo/pull/684#issuecomment-2645408352
But presumably not in an isolated test case 🤔